ompi/mca/coll/acoll/coll_acoll_allreduce.c

```diff
@@ -0,0 +1,553 @@
+                           (char *) data->xpmem_raddr[0] + chunk * l2_local_rank * dsize,
+                           my_count_size, dtype);
+        }
+    }
+
+    err = ompi_coll_base_barrier_intra_tree(comm, module);
+    if (!subc->xpmem_use_sr_buf) {
+        memcpy(rbuf, tmp_rbuf, total_dsize);
+    }
+    return err;
+}
+
+static inline int mca_coll_acoll_allreduce_xpmem_f(const void *sbuf, void *rbuf, int count,
+                                                   struct ompi_datatype_t *dtype,
+                                                   struct ompi_op_t *op,
+                                                   struct ompi_communicator_t *comm,
+                                                   mca_coll_base_module_t *module)
+{
+    int size;
+    size_t total_dsize, dsize;
+    mca_coll_acoll_module_t *acoll_module = (mca_coll_acoll_module_t *) module;
+
+    coll_acoll_subcomms_t *subc;
+    int cid = ompi_comm_get_local_cid(comm);
+    subc = &acoll_module->subc[cid];
+    coll_acoll_init(module, comm, subc->data);
+    coll_acoll_data_t *data = subc->data;
+    if (NULL == data) {
+        return -1;
+    }
+
+    size = ompi_comm_size(comm);
+    ompi_datatype_type_size(dtype, &dsize);
+    total_dsize = dsize * (unsigned long) count;
+
+    char *tmp_sbuf = NULL;
+    char *tmp_rbuf = NULL;
+    if (!subc->xpmem_use_sr_buf) {
+        tmp_rbuf = (char *) data->scratch;
+        tmp_sbuf = (char *) data->scratch + (subc->xpmem_buf_size) / 2;
+        if ((sbuf == MPI_IN_PLACE)) {
+            memcpy(tmp_sbuf, rbuf, total_dsize);
+        } else {
+            memcpy(tmp_sbuf, sbuf, total_dsize);
+        }
+    } else {
+        tmp_sbuf = (char *) sbuf;
+        tmp_rbuf = (char *) rbuf;
+        if (sbuf == MPI_IN_PLACE) {
+            tmp_sbuf = (char *) rbuf;
+        }
+    }
+    void *sbuf_vaddr[1] = {tmp_sbuf};
+    void *rbuf_vaddr[1] = {tmp_rbuf};
+    int err = MPI_SUCCESS;
+    int rank = ompi_comm_rank(comm);
+
+    err = comm->c_coll->coll_allgather(sbuf_vaddr, sizeof(void *), MPI_BYTE, data->allshm_sbuf,
+                                       sizeof(void *), MPI_BYTE, comm,
+                                       comm->c_coll->coll_allgather_module);
+    if (err != MPI_SUCCESS) {
+        return err;
+    }
+    err = comm->c_coll->coll_allgather(rbuf_vaddr, sizeof(void *), MPI_BYTE, data->allshm_rbuf,
+                                       sizeof(void *), MPI_BYTE, comm,
+                                       comm->c_coll->coll_allgather_module);
+
+    if (err != MPI_SUCCESS) {
+        return err;
+    }
+
+    register_and_cache(size, total_dsize, rank, data);
+
+    int chunk = count / size;
+    int my_count_size = (rank == (size - 1)) ? (count / size) + count % size : count / size;
+    if (rank == 0) {
+        if (sbuf != MPI_IN_PLACE)
+            memcpy(tmp_rbuf, sbuf, my_count_size * dsize);
+    } else {
+        ompi_3buff_op_reduce(op, (char *) data->xpmem_saddr[0] + chunk * rank * dsize,
```

Comment:
is the 3 operator reduce function to maintain the order?

Reply:
I think this was a bit faster than just copying the chunk first and then reducing later in the following "for" loop.